service/lambda/api_op_AddLayerVersionPermission.go

```diff
@@ -208,6 +209,52 @@ func (m *opAddLayerVersionPermissionResolveEndpointMiddleware) HandleSerialize(c
 		)
 	}
 
+	authSchemes, err := internalauth.GetAuthenticationSchemes(&resolvedEndpoint.Properties)
+	if err != nil || len(authSchemes) == 0 {
+		return out, metadata, fmt.Errorf("Failed to resolve authentication scheme")
+	}
+
+	supportedAuthSchemeFound := false
+	for _, authScheme := range authSchemes {
+		name := authScheme.GetName()
+		_, supportedAuthSchemeFound = internalauth.SupportedSchemes[name]
+		if supportedAuthSchemeFound {
```

Comment:
**nit/style**:

We don't really need the name (or a set of supported auth schemes for that matter) if we are going to do a type assertion over the same set.

e.g.


```go
switch v := authScheme.(type) {
    case *internalauth.AuthenticationSchemeV4:
        if v.SigningName != nil {
            ctx = awsmiddleware.SetSigningName(ctx, v.SigningName)
        }
        if v.SigningRegion != nil {
            ctx = awsmiddleware.SetSigningRegion(ctx, v.SigningRegion)
        }
    case *internalauth.AuthenticationSchemeV4A:
}
```


Alternatively you could just move some of this into the runtime and let each auth scheme handle updating the context:

```go
interface AuthenticationScheme {
    fun SetAuthContext(context.Context): context.Context
}

func (s *AuthenticationSchemeV4) SetAuthContext(ctx context.Context): context.Context {
        if s.SigningName != nil {
            ctx = awsmiddleware.SetSigningName(ctx, s.SigningName)
        }

        if s.SigningRegion != nil {
            ctx = awsmiddleware.SetSigningRegion(ctx, s.SigningRegion)
        }

        return ctx
}
```

Reply:
i do like this. and attempted this. but there were a couple of hiccups:
1. there is a s3 specific context setter `s3cust.SetSignerVersion` that i would need to import into this package. and doing so, would result in an import cycle: s3cust package imports a v4a package which imports this package (in order to use the context getters i define in this `internal/auth` package)
2. the signing name and signing region defaults are service specific, so a shared library function like `SetAuthContext` wouldnt be as clean as taking in a single `ctx` param. and will potentially get messy with the fact that sigv4 and sigv4a use different data structures to model signing region (`string` vs `[]string`).

given, that this endpoint-based auth scheme resolution code is likely to be replace with the auth scheme resolution refactor work in the future. im not sure if much value is gained in style optimizing it.

Reply:
> given, that this endpoint-based auth scheme resolution code is likely to be replaced with the auth scheme resolution refactor work in the future. im not sure if much value is gained in style optimizing it.

Endpoints will still be allowed to influence signing (they just won't eventually be setting e.g. signer version). Let's ignore that for now though to ship EP2.0.

> there is a s3 specific context setter s3cust.SetSignerVersion that i would need to import into this package. and doing so, would result in an import cycle: s3cust package imports a v4a package which imports this package (in order to use the context getters i define in this internal/auth package)

It looks like we are duplicating this logic in [event bridge](https://github.com/isaiahvita/aws-sdk-go-v2/blob/feat-ep20-authscheme/service/eventbridge/internal/customizations/signer_wrapper.go#L28)

> the signing name and signing region defaults are service specific, so a shared library function like SetAuthContext wouldnt be as clean as taking in a single ctx param. and will potentially get messy with the fact that sigv4 and sigv4a use different data structures to model signing region (string vs []string)

I'm not sure I understand this, the function would be a method on the authentication scheme allowing each to set context properties appropriate for the scheme.

Reply:
> It looks like we are duplicating this logic in [event bridge](https://github.com/isaiahvita/aws-sdk-go-v2/blob/feat-ep20-authscheme/service/eventbridge/internal/customizations/signer_wrapper.go#L28)
yah thats correct. if you `git diff` that file `signer_wrapper` in `s3` and `eventbridge`, theyre identical.

Reply:
in any case, i got rid of the initial nit style ask which was to remove `GetName`. so resolving this thread.